orttraining/orttraining/core/optimizer/gist_encode_decode.cc

```diff
@@ -45,74 +45,115 @@ static std::vector<GraphEdgeHelper> GetNodeOutputEdges(const Node& node) {
-          graph.AddEdge(input_edges_dst[i].src_node, decode.Index(), input_edges_dst[i].src_arg_index, 0);
-          late_decoding = true;
+    // Override compression_type for lossless compression case(s) (eg. bool -> Pack1)
+    ONNX_NAMESPACE::DataType type_string = curr_node_output_arg->Type();
+    if (*type_string == "bool" || *type_string == "tensor(bool)") {
+      std::cout << "(Lossless) override compression type to Pack1 for tensor: " << curr_node_output_arg->Name() << std::endl;
+      compression_type = "GistPack1";
+    }
+    
+    if (compression_type == "GistPack1" || compression_type == "GistPack8" || compression_type == "GistPackMsfp15") {
+      compressed_tensor.mutable_tensor_type()->set_elem_type(ONNX_NAMESPACE::TensorProto_DataType_UINT8);
+    }
+    else if (compression_type == "GistPack16") {
+      compressed_tensor.mutable_tensor_type()->set_elem_type(ONNX_NAMESPACE::TensorProto_DataType_FLOAT16);
+    }
+    else {
+      assert (0); // "Gist compression type not supported"
+    }
+    
+    bool tensor_size_compressed = false;
+    for (int i = 0; i < curr_node_output_arg->Shape()->dim_size(); i++) {
+      if (curr_node_output_arg->Shape()->dim(i).dim_value() != 0) {
+        if (compression_type == "GistPack1" && !tensor_size_compressed) {
+          compressed_tensor.mutable_tensor_type()->mutable_shape()->add_dim()->set_dim_value((curr_node_output_arg->Shape()->dim(i).dim_value()  + GIST_PACK1_FACTOR -1)/GIST_PACK1_FACTOR); 
+          tensor_size_compressed = true;
         }
-        i++;
+        else {
+          compressed_tensor.mutable_tensor_type()->mutable_shape()->add_dim()->set_dim_value(curr_node_output_arg->Shape()->dim(i).dim_value());
+        } 
       }
-    } else if (!early_encoding) {
-      graph.AddEdge(encode.Index(), output_edge.dst_node, 0, output_edge.dst_arg_index);
-      early_encoding = true;
+      else {
+        compressed_tensor.mutable_tensor_type()->mutable_shape()->add_dim()->set_dim_param(curr_node_output_arg->Shape()->dim(i).dim_param());
+      }
+    }
+
+    // Create encode/decode nodes
+    std::string gist_pair_name = graph.GenerateNodeName(GIST_PAIR_NODE_NAME_BASE);
+
+    std::string encode_node_name = "encode_" + gist_pair_name;
+    auto& encode_output_def_compressed_arg = graph.GetOrCreateNodeArg("compr_" + curr_node_output_arg->Name(), &compressed_tensor);
+    auto& encode = graph.AddNode(encode_node_name, compression_type + "Encoder", "Encode", {curr_node_output_arg}, {&encode_output_def_compressed_arg});
+    encode.SetPriority(static_cast<int>(ExecutionPriority::LOCAL_HIGH));
+
+    std::string decode_node_name = "decode_" + gist_pair_name;
+    auto& decode_output_def_uncompressed_arg = graph.GetOrCreateNodeArg("uncompr_" + curr_node_output_arg->Name(), curr_node_output_arg->TypeAsProto());
+    int curr_dec_priority = graph.GenerateDecodePriority();
+    assert(curr_dec_priority > 0);
```

Comment:
Related to the above comment about `Graph::GenerateDecodePriority()`... could you at least add a comment here to explain how the auto-decrementing priority is used here?

(From reading this method, it's not clear why the priority couldn't be a local variable in this method - or at worst an instance variable in this class. What assumptions are being made about multiple calls to the method, and what state in the graph is preserved across these calls?)

Reply:
Added a comment on why decoder priority is generated using the auto-decrementing done in `GenerateDecodePriority()` function. Also, added comment to clarify why encoder priority is kept same across all encoders.

`priority_generator_` is now a member variable in this class. Subsequently, using this member variable it is ensured that across multiple calls to ```AddEncodeDecode``` method, priority for decoder used is unique.